channeldb/migration29/migration.go

```diff
@@ -0,0 +1,498 @@
+	log.Infof("Migrating revocation logs, might take a while...")
+
+	var (
+		err error
+
+		// finished is used to exit the for loop.
+		finished bool
+
+		// total is the number of total un-migrated records.
+		total uint64
+
+		// migrateCount is the total records migrated in current
+		// migration.
+		migrateCount float64
+	)
+
+	// First of all, read the number of un-migrated records.
+	err = kvdb.Update(db, func(tx kvdb.RwTx) error {
+		total, err = fetchTotalNumOfRecords(tx)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, func() {})
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Total un-migrated logs: %d", total)
+
+	for {
+		if finished {
+			log.Infof("Migrating revocation logs finished!")
+			break
+		}
+
+		// Process the migration.
+		err = kvdb.Update(db, func(tx kvdb.RwTx) error {
+			finished, err = processMigration(tx)
+			if err != nil {
+				return err
+			}
+			return nil
+		}, func() {})
+		if err != nil {
+			return err
+		}
+
+		// Eeach time we finished the above process, we've processed at
+		// most `recordsPerTx` records.
+		migrateCount += recordsPerTx
+
+		// Calculate and log the progress if the progress is less than
+		// one.
+		progress := migrateCount / float64(total) * 100
+		if progress >= 100 {
+			continue
+		}
+
+		log.Infof("Migration progress: %.3f%%, still have: %d",
+			progress, total-uint64(migrateCount))
+	}
+
+	// Once the migration completes, we can now safety delete the old
+	// revocation logs.
+	// TODO: do this in a follow up migration?
+	if err := deleteOldBuckets(db); err != nil {
+		return fmt.Errorf("deleteOldBuckets err: %v", err)
+	}
+
+	log.Info("Old revocation log buckets removed!")
+	return nil
+}
+
+// processMigration finds the next un-migrated revocation logs, reads a max
+// number of `recordsPerTx` records, converts them into the new revocation logs
+// and save them to disk.
+func processMigration(tx kvdb.RwTx) (bool, error) {
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if openChanBucket == nil {
+		return false, fmt.Errorf("root bucket not found")
+	}
+
+	// Locate the next migration height.
+	locator, err := locateNextUpdateNum(openChanBucket)
+	if err != nil {
+		return false, fmt.Errorf("locator got error: %v", err)
+	}
+
+	// If the returned locator is nil, we've done migrating the logs.
+	if locator == nil {
+		return true, nil
+	}
+
+	// Read a list of old revocation logs.
+	entryMap, err := readOldRevocationLogs(openChanBucket, locator)
+	if err != nil {
+		return false, fmt.Errorf("read old logs err: %v", err)
+	}
+
+	// Migrate the revocation logs.
+	return false, writeRevocationLogs(openChanBucket, entryMap)
+}
+
+// deleteOldBuckets iterates all the channel buckets and deletes the old
+// revocation buckets.
+func deleteOldBuckets(db kvdb.Backend) error {
+	// locators records all the chan buckets found in the database.
+	locators := make([]*updateLocator, 0)
```

Comment:
Can be slightly simplified to: `var locators []*updateLoactor`.

Reply:
Done.